File: valid-anagram/anagram.py
# Valid Anagram

# Brute force
# def isAnagram(s, t):
#     return sorted(s) == sorted(t) # same -> true

# Optimal
def isAnagram(s,t):
    if len(s) != len(t):
        return False
    
    count = {}
    
    for char in s: # Increment
        count[char] = count.get(char, 0) + 1
        
    for char in t: # decrement
        count[char] = count.get(char, 0) - 1
        if count[char] < 0: # extra charcter
            return False
        
    return True

# Test 1
s = "east"
t = "seat"
print(isAnagram(s,t))

File: valid-anagram/anagram.js
// Valid Anagram

// Brute Force
// function isAnagram(s, t) {
//     return s.split('').sort().join('') === t.split('').sort().join('') // true
// }

// Optimal Solution
function isAnagram (s, t) {
    if (s.length !== t.length) {
        return false
    }

    const count = {}

    for (let char of s) {
        // increment
        count[char] = (count[char] || 0) + 1;
    }

    for (let char of t) {
        // decrement
        count[char] = (count[char] || 0) - 1;
        if(count[char] < 0) {
            return false
        }
    }

    return true;
}

const s = "east";
const t = "seat";

console.log(isAnagram(s, t))